File: teaching/anthropic/anthropic-protocol-book/code/python/ch12/mcp_tasks.py
"""
Chapter 12: MCP Tasks (SEP-1686).

Implements the MCP Tasks specification for asynchronous task execution.
Provides task submission, status polling, result retrieval, listing,
and cancellation. Designed to work with any MCPClient.

MCP Tasks Spec: modelcontextprotocol.io/specification/2025-11-25/basic/utilities/tasks
SEP-1686: modelcontextprotocol.io/seps/1686-tasks
SEP-2663: modelcontextprotocol.io/seps/2663-tasks-extension
"""

from __future__ import annotations

import asyncio
import time
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from mcp_client import MCPClient


# ──────────────────────────────────────────────────────────────────────
# Task status constants
# ──────────────────────────────────────────────────────────────────────

TASK_STATUS_WORKING = "working"
TASK_STATUS_INPUT_REQUIRED = "input_required"
TASK_STATUS_COMPLETED = "completed"
TASK_STATUS_FAILED = "failed"
TASK_STATUS_CANCELLED = "cancelled"

TERMINAL_STATUSES = {TASK_STATUS_COMPLETED, TASK_STATUS_FAILED, TASK_STATUS_CANCELLED}


# ──────────────────────────────────────────────────────────────────────
# MCPTask
# ──────────────────────────────────────────────────────────────────────


class MCPTask:
    """Manages the lifecycle of an MCP Task.

    Wraps the SEP-1686 task operations: submission, status polling,
    result retrieval, listing, and cancellation.

    Usage:
        client = MCPClient()
        await client.connect_stdio("python", ["mcp_server.py"])
        task_mgr = MCPTask(client)
        task_result = await task_mgr.submit("analyze_data", {"dataset": "q4"})
        print(task_result["task"]["taskId"])
        result = await task_mgr.wait_for_completion("task-id-here")
    """

    def __init__(self, client: "MCPClient"):
        self._client = client
        self._tasks: Dict[str, dict] = {}

    # ── Task submission ──────────────────────────────────────────

    async def submit_task(
        self, tool_name: str, arguments: Dict[str, Any], ttl: Optional[int] = None
    ) -> str:
        """Submit a task-augmented tool call and return the task ID.

        Args:
            tool_name: Name of the tool to invoke.
            arguments: Tool arguments.
            ttl: Task time-to-live in milliseconds.

        Returns:
            The task ID assigned by the server.

        Raises:
            RuntimeError: If the server does not support task-augmented requests.
        """
        params: Dict[str, Any] = {
            "name": tool_name,
            "arguments": arguments,
            "task": {},
        }
        if ttl is not None:
            params["task"]["ttl"] = ttl

        response = await self._client.send_request("tools/call", params)

        task_data = response.get("task")
        if not task_data:
            raise RuntimeError(
                "Server did not return task data. "
                "Does the server support task-augmented tools/call?"
            )

        task_id = task_data["taskId"]
        self._tasks[task_id] = task_data
        return task_id

    # ── Status polling ───────────────────────────────────────────

    async def get_task_status(self, task_id: str) -> dict:
        """Get the current status of a task.

        Returns the full Task object as defined by the spec:
        {taskId, status, statusMessage, createdAt, lastUpdatedAt, ttl, pollInterval}
        """
        result = await self._client.send_request("tasks/get", {"taskId": task_id})
        self._tasks[task_id] = result
        return result

    async def wait_for_completion(
        self, task_id: str, timeout: float = 60.0, poll_interval: Optional[float] = None
    ) -> dict:
        """Poll for task completion and return the result.

        Args:
            task_id: The task ID to wait for.
            timeout: Maximum time to wait in seconds.
            poll_interval: Optional override for polling interval.
                          If None, uses the server's suggested pollInterval.

        Returns:
            The task result on success. On timeout, raises TimeoutError.
            On failure, returns {"success": False, "error": "..."}.

        Raises:
            TimeoutError: If task doesn't complete within timeout.
        """
        start = time.monotonic()
        default_interval = 1.0

        while True:
            elapsed = time.monotonic() - start
            if elapsed >= timeout:
                raise TimeoutError(
                    f"Task {task_id} did not complete within {timeout}s"
                )

            status = await self.get_task_status(task_id)
            current_status = status.get("status", "")

            if current_status in TERMINAL_STATUSES:
                if current_status == TASK_STATUS_COMPLETED:
                    result = await self.get_task_result(task_id)
                    return {"success": True, "result": result}
                elif current_status == TASK_STATUS_FAILED:
                    return {
                        "success": False,
                        "error": status.get("statusMessage", "Task failed"),
                    }
                elif current_status == TASK_STATUS_CANCELLED:
                    return {
                        "success": False,
                        "error": "Task was cancelled",
                    }

            interval = poll_interval or status.get("pollInterval", default_interval)
            # Convert ms to seconds
            if interval > 1:
                interval = interval / 1000.0
            interval = min(interval, timeout - elapsed)

            await asyncio.sleep(interval)

    # ── Result retrieval ─────────────────────────────────────────

    async def get_task_result(self, task_id: str) -> dict:
        """Retrieve the result of a completed task.

        Blocks until the task reaches a terminal status.
        Returns the actual operation result (e.g., CallToolResult).
        """
        result = await self._client.send_request("tasks/result", {"taskId": task_id})
        return result

    # ── Task listing ─────────────────────────────────────────────

    async def list_tasks(self, cursor: Optional[str] = None) -> dict:
        """List all tasks managed by the server.

        Returns: {tasks: [...], nextCursor: str|None}
        """
        params: Dict[str, Any] = {}
        if cursor:
            params["cursor"] = cursor
        return await self._client.send_request("tasks/list", params)

    # ── Task cancellation ────────────────────────────────────────

    async def cancel_task(self, task_id: str) -> dict:
        """Cancel a running task.

        Returns the task object with status "cancelled".

        Raises:
            RuntimeError: If the task is already in a terminal status.
        """
        return await self._client.send_request("tasks/cancel", {"taskId": task_id})

    # ── High-level helpers ───────────────────────────────────────

    async def submit_and_wait(
        self,
        tool_name: str,
        arguments: Dict[str, Any],
        ttl: Optional[int] = None,
        timeout: float = 60.0,
    ) -> dict:
        """Convenience: submit a task and wait for completion.

        Returns: {"success": True, "result": ...} or {"success": False, "error": ...}
        """
        task_id = await self.submit_task(tool_name, arguments, ttl=ttl)
        return await self.wait_for_completion(task_id, timeout=timeout)

    async def poll_until(
        self,
        task_id: str,
        predicate: callable,
        timeout: float = 60.0,
        poll_interval: float = 1.0,
    ) -> dict:
        """Poll until a custom predicate is satisfied.

        Args:
            task_id: Task ID to poll.
            predicate: Callable(status_dict) -> bool. Return True to stop.
            timeout: Maximum time in seconds.
            poll_interval: Seconds between polls.

        Returns:
            The final status dict when predicate returns True.
        """
        start = time.monotonic()
        while True:
            elapsed = time.monotonic() - start
            if elapsed >= timeout:
                raise TimeoutError(f"Polling timeout after {timeout}s")

            status = await self.get_task_status(task_id)
            if predicate(status):
                return status

            await asyncio.sleep(poll_interval)

File: teaching/anthropic/anthropic-protocol-book/code/node/ch12/MCPTasks.ts
/**
 * Chapter 12: MCP Tasks (SEP-1686) - TypeScript.
 *
 * Implements the MCP Tasks specification for asynchronous task execution.
 * Provides task submission, status polling, result retrieval, listing,
 * and cancellation. Designed to work with MCPClient.
 *
 * MCP Tasks Spec: modelcontextprotocol.io/specification/2025-11-25/basic/utilities/tasks
 * SEP-1686: modelcontextprotocol.io/seps/1686-tasks
 * SEP-2663: modelcontextprotocol.io/seps/2663-tasks-extension
 */

import type { MCPClient } from "./MCPClient";

// ────────────────────────────────────────────────────────────────
// Constants
// ────────────────────────────────────────────────────────────────

export const TASK_STATUS_WORKING = "working";
export const TASK_STATUS_INPUT_REQUIRED = "input_required";
export const TASK_STATUS_COMPLETED = "completed";
export const TASK_STATUS_FAILED = "failed";
export const TASK_STATUS_CANCELLED = "cancelled";

export const TERMINAL_STATUSES = new Set([
  TASK_STATUS_COMPLETED,
  TASK_STATUS_FAILED,
  TASK_STATUS_CANCELLED,
]);

export interface TaskStatus {
  taskId: string;
  status: string;
  statusMessage?: string;
  createdAt: string;
  lastUpdatedAt: string;
  ttl: number | null;
  pollInterval?: number;
}

export interface TaskResult {
  success: boolean;
  result?: Record<string, unknown>;
  error?: string;
}

// ────────────────────────────────────────────────────────────────
// MCPTask
// ────────────────────────────────────────────────────────────────

export class MCPTask {
  private client: MCPClient;
  private tasks = new Map<string, TaskStatus>();

  constructor(client: MCPClient) {
    this.client = client;
  }

  // ── Task submission ────────────────────────────────────

  async submitTask(
    toolName: string,
    args: Record<string, unknown>,
    ttl?: number
  ): Promise<string> {
    const params: Record<string, unknown> = {
      name: toolName,
      arguments: args,
      task: {} as Record<string, unknown>,
    };
    if (ttl !== undefined) {
      (params.task as Record<string, unknown>).ttl = ttl;
    }

    const response = (await this.client.sendRequest("tools/call", params)) as Record<string, unknown>;
    const taskData = response.task as Record<string, unknown> | undefined;

    if (!taskData || !taskData.taskId) {
      throw new Error(
        "Server did not return task data. Does the server support task-augmented tools/call?"
      );
    }

    const taskId = taskData.taskId as string;
    this.tasks.set(taskId, taskData as unknown as TaskStatus);
    return taskId;
  }

  // ── Status polling ─────────────────────────────────────

  async getTaskStatus(taskId: string): Promise<TaskStatus> {
    const result = (await this.client.sendRequest("tasks/get", {
      taskId,
    })) as TaskStatus;
    this.tasks.set(taskId, result);
    return result;
  }

  async waitForCompletion(
    taskId: string,
    timeout = 60.0,
    pollInterval?: number
  ): Promise<TaskResult> {
    const start = Date.now();
    const defaultInterval = 1.0;

    while (true) {
      const elapsed = (Date.now() - start) / 1000;
      if (elapsed >= timeout) {
        throw new Error(`Task ${taskId} did not complete within ${timeout}s`);
      }

      const status = await this.getTaskStatus(taskId);
      const currentStatus = status.status;

      if (TERMINAL_STATUSES.has(currentStatus)) {
        if (currentStatus === TASK_STATUS_COMPLETED) {
          const result = await this.getTaskResult(taskId);
          return { success: true, result: result as Record<string, unknown> };
        } else if (currentStatus === TASK_STATUS_FAILED) {
          return {
            success: false,
            error: status.statusMessage ?? "Task failed",
          };
        } else if (currentStatus === TASK_STATUS_CANCELLED) {
          return { success: false, error: "Task was cancelled" };
        }
      }

      let interval = pollInterval ?? status.pollInterval ?? defaultInterval;
      // Convert ms to seconds if value is large
      if (interval > 1) interval = interval / 1000;
      interval = Math.min(interval, timeout - elapsed);

      await new Promise((resolve) => setTimeout(resolve, interval * 1000));
    }
  }

  // ── Result retrieval ───────────────────────────────────

  async getTaskResult(taskId: string): Promise<Record<string, unknown>> {
    return (await this.client.sendRequest("tasks/result", {
      taskId,
    })) as Record<string, unknown>;
  }

  // ── Task listing ───────────────────────────────────────

  async listTasks(cursor?: string): Promise<{ tasks: TaskStatus[]; nextCursor: string | null }> {
    const params: Record<string, unknown> = {};
    if (cursor) params.cursor = cursor;
    return (await this.client.sendRequest("tasks/list", params)) as {
      tasks: TaskStatus[];
      nextCursor: string | null;
    };
  }

  // ── Task cancellation ──────────────────────────────────

  async cancelTask(taskId: string): Promise<TaskStatus> {
    return (await this.client.sendRequest("tasks/cancel", {
      taskId,
    })) as TaskStatus;
  }

  // ── High-level helpers ─────────────────────────────────

  async submitAndWait(
    toolName: string,
    args: Record<string, unknown>,
    options?: { ttl?: number; timeout?: number }
  ): Promise<TaskResult> {
    const taskId = await this.submitTask(toolName, args, options?.ttl);
    return await this.waitForCompletion(taskId, options?.timeout ?? 60.0);
  }

  async pollUntil(
    taskId: string,
    predicate: (status: TaskStatus) => boolean,
    timeout = 60.0,
    pollInterval = 1.0
  ): Promise<TaskStatus> {
    const start = Date.now();

    while (true) {
      const elapsed = (Date.now() - start) / 1000;
      if (elapsed >= timeout) {
        throw new Error(`Polling timeout after ${timeout}s`);
      }

      const status = await this.getTaskStatus(taskId);
      if (predicate(status)) {
        return status;
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval * 1000));
    }
  }
}
